### ISSUE TITLE
[QUESTION] Issue with Adding a Custom Transformer Block in Megatron

### ISSUE THREAD

**Issue by BhagawanPanditi** [NVIDIA/Megatron-LM#1436]
I'm implementing Multi-Token Prediction (similar to DeepSeek V3) in the Megatron framework. The goal is to predict one extra token beyond the next token. This requires an **extra transformer block** that is separate from the main transformer stack. The extra block:
1. Takes the output of the main stack.
2. Concatenates it with input embeddings for the extra token.
3. Projects this concatenated tensor as input to the extra block.
4. Computes a new loss, which is added to the original main loss.

I have already modified the necessary files to get input and labels for this extra block. However, I'm facing issues with implementing the extra transformer block.

#### **First Approach:**
I attempted to modify the config's `num_layers` temporarily to create a single-layer transformer block:

```python
original_num_layers = self.config.num_layers  # Store original value
self.config.num_layers = 1  # Temporarily set to 1
self.mtp_decoder = TransformerBlock(
    config=self.config,
    spec=transformer_layer_spec,
    pre_process=self.pre_process,
    post_process=self.post_process,
)
self.config.num_layers = original_num_layers  # Restore original value
```

This resulted in a **checkpoint saving error**:

```
File “/opt/megatron-lm/megatron/core/dist_checkpointing/validation.py”, line 470, in _validate_sharding_for_key
    raise CheckpointingException(
megatron.core.dist_checkpointing.core.CheckpointingException: Invalid access pattern for ShardedTensor(key='optimizer.state.fp32_param.model.mtp_decoder.layers.self_attention.linear_proj.weight', dtype=torch.float32, local_shape=(768, 768), global_shape=(12, 768, 768), global_offset=(0, 0, 0), axis_fragmentations=(12, 1, 1), replica_id=(0, 0, 0), prepend_axis_num=1, allow_shape_mismatch=False, flattened_range=None): tensor([[[1]],
        [[0]],
        [[0]],
        [[0]],
        [[0]],
        [[0]],
        [[0]],
        [[0]],
        [[0]],
        [[0]],
        [[0]],
        [[0]]], dtype=torch.int32)
```

I suspect this happens because the original stack has 12 layers, and restoring `num_layers` after creating the extra block leads to issues.

#### **Second Approach:**
I tried using a deep copy of the config:

```python
self.mtp_config = copy.deepcopy(self.config)
self.mtp_config.num_layers = 1
self.mtp_decoder = TransformerBlock(
    config=self.mtp_config,
    spec=transformer_layer_spec,
    pre_process=self.pre_process,
    post_process=self.post_process,
)
```

This led to an **NCCL/CUDA API hang**:

```
[rank0]:[E226 14:50:58.064411450 ProcessGroupNCCL.cpp:1496] [PG ID 5 PG GUID 133 Rank 0] ProcessGroupNCCL's watchdog got stuck for 480 seconds without making progress in monitoring enqueued collectives. This typically indicates a NCCL/CUDA API (e.g., CudaEventDestroy) hang blocking the watchdog, and could be triggered by another thread holding the GIL inside a CUDA API (for example, CudaEventDestroy), or other deadlock-prone behaviors.
```

### **Current Implementation**
**File path:** `/opt/megatron-lm/megatron/core/models/gpt/gpt_model.py`
#### **Modifications in `GPTModel` Class**

```python
# Multi-token prediction processing
self.mtp_config = copy.deepcopy(self.config)
self.mtp_config.num_layers = 1
self.mtp_decoder = TransformerBlock(
    config=self.mtp_config,
    spec=transformer_layer_spec,
    pre_process=self.pre_process,
    post_process=self.post_process,
)

self.hidden_norm = TENorm(
    config=self.config,
    hidden_size=self.config.hidden_size,
    eps=self.config.layernorm_epsilon
)
self.embedding_norm = TENorm(
    config=self.config,
    hidden_size=self.config.hidden_size,
    eps=self.config.layernorm_epsilon
)
self.projection = TERowParallelLinear(
    input_size=2 * self.config.hidden_size,
    output_size=self.config.hidden_size,
    config=self.config,
    init_method=self.config.init_method,
    bias=True,
    input_is_parallel=True,
    skip_bias_add=False,
    is_expert=False
)
```

#### **Forward Method Modifications**

```python
mtp_logits = None
if mtp_input_ids is not None and self.pre_process:
    # Get embeddings for MTP input
    mtp_decoder_embeddings = self.embedding(
        input_ids=mtp_input_ids,
        position_ids=mtp_position_ids
    )

    # Apply normalization
    normalized_hidden = self.hidden_norm(hidden_states)
    normalized_embeddings = self.embedding_norm(mtp_decoder_embeddings)
    
    # Concatenate along hidden dimension
    concatenated = torch.cat(
        [normalized_hidden, normalized_embeddings],
        dim=-1
    )  # Shape: [seq_len, batch_size, 2*hidden_size]

    mtp_decoder_input, _ = self.projection(concatenated)
    
    # Pass through single transformer block
    mtp_hidden_states = self.mtp_decoder(
        hidden_states=mtp_decoder_input,
        attention_mask=mtp_attention_mask,
        inference_params=inference_params,
        rotary_pos_emb=rotary_pos_emb,
        rotary_pos_cos=rotary_pos_cos,
        rotary_pos_sin=rotary_pos_sin,
        packed_seq_params=packed_seq_params,
        **(extra_block_kwargs or {}),
    )
    
    # Get MTP logits using shared output layer
    output_weight = None
    if self.share_embeddings_and_output_weights:
        output_weight = self.shared_embedding_or_output_weight()
    mtp_logits, _ = self.output_layer(
        mtp_hidden_states,
        weight=output_weight,
        runtime_gather_output=runtime_gather_output
    )
```

#### **Loss Computation Modifications**

```python
mtp_loss = None
if mtp_labels is not None and mtp_logits is not None:
    mtp_loss = self.compute_language_model_loss(mtp_labels, mtp_logits)

if mtp_loss is not None:
    combined_loss = loss + self.config.mtp_weight * mtp_loss
    return combined_loss
```

### **Questions**
1. Is there a proper way to create an independent single-layer transformer block within the `GPTModel` class that does not interfere with checkpointing or distributed training?
2. Are there specific considerations for adding custom components to Megatron models that I might be missing?

Any guidance would be helpful!

**Comment by shifangx** [comment]
Multi-Token Prediction (MTP) is supported now in MCore.
Maybe you can use this implement as a reference.

**Comment by sbhavani** [comment]
Closing as resolved - MTP is now officially supported in Megatron Core.
